[FIO internal] fiohab: remove safeguard for incorrect rpmb key

ricardosalveti opened the pull request:
The initial check was done to cover cases in which the RPMB key is
already burned (e.g. test keys), to block the close process as that
would result in OP-TEE using the HUK-derived RPMB key instead.

This causes a few problems:
- Will not work on devices without eMMC/RPMB
- Can end up burning the wrong RPMB key (bugs in OP-TEE)
- Always cause a panic during the closing process

Remove as fusing shouldn't be blocked by the RPMB state.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ldts commented:
1) issue 1:  um weird, the first thing it does is to call init_mmc_device. if there is no mmc the call will not proceed. how come uboot thinks there is mmc?

2) issue 2: um weird as well: how can this code burnt any keys? if that was the case all boards would be bricked immediately...so no, I dont think this is an issue either.

3) issue 3: always causes a panic during the closing process: why? is there a bug in optee? it is true that optee tends to panic for stuff instead of returning errors...

I obviously disagree: fusing should be blocked by the RPMB state (unless we want to brick boards :) ) - once closed, RPMB will no longer be accesible if it was already provisioned which is not a desirable state.

I think this PR needs further investigation. But if you want something quick and dirty  why dont you just compile it out with CONFIG_MMC?

ricardosalveti commented:
> 1. issue 1:  um weird, the first thing it does is to call init_mmc_device. if there is no mmc the call will not proceed. how come uboot thinks there is mmc?

We can have boards with sdcard but not emmc (.e.g imx6ullevk), and I believe that find_mmc_device will still be valid when sdcard is found, right?

> 2. issue 2: um weird as well: how can this code burnt any keys? if that was the case all boards would be bricked immediately...so no, I dont think this is an issue either.

To check for rpmb key this code was performing write_persistent_value, which can end up causing optee to burn the key (e.g. cases where huk checks are incorrect).

> 3. issue 3: always causes a panic during the closing process: why? is there a bug in optee? it is true that optee tends to panic for stuff instead of returning errors...

```
I/TC: RPMB: Using generated key
E/TC:? 0 tee_rpmb_verify_key_sync_counter:997 Verify key returning 0xffff0008
E/TC:? 0 tee_rpmb_verify_key_sync_counter:997 Verify key returning 0xffff0008
E/TC:? 0
E/TC:? 0 TA panicked with code 0xffff0007
E/LD:  Status of TA 22250a54-0bf1-48fe-8002-7b20f1c9c9b1
E/LD:   arch: arm
E/LD:  region  0: va 0x00102000 pa 0x9e100000 size 0x002000 flags rw-s (ldelf)
E/LD:  region  1: va 0x00104000 pa 0x9e102000 size 0x006000 flags r-xs (ldelf)
E/LD:  region  2: va 0x0010a000 pa 0x9e108000 size 0x001000 flags rw-s (ldelf)
E/LD:  region  3: va 0x0010b000 pa 0x9e109000 size 0x004000 flags rw-s (ldelf)
E/LD:  region  4: va 0x0010f000 pa 0x9e10d000 size 0x001000 flags r--s
E/LD:  region  5: va 0x00110000 pa 0x9e11f000 size 0x005000 flags rw-s (stack)
E/LD:  region  6: va 0x00120000 pa 0x00001000 size 0x009000 flags r-xs [0]
E/LD:  region  7: va 0x00129000 pa 0x0000a000 size 0x008000 flags rw-s [0]
E/LD:  region  8: va 0x00200000 pa 0x9cf91c98 size 0x001000 flags rw-- (param)
E/LD:  region  9: va 0x00201000 pa 0x9cf91c68 size 0x001000 flags rw-- (param)
E/LD:   [0] 22250a54-0bf1-48fe-8002-7b20f1c9c9b1 @ 0x00120000
E/LD:  Call stack:
E/LD:   0x0012188c
E/LD:   0x001214f3
E/LD:   0x001201cd
E/LD:   0x0012218d
E/LD:   0x001266ef
E/LD:   0x0012022c
```

Have to check, but I believe this is because of the new huk checks when verifying the rpmb key.

> I obviously disagree: fusing should be blocked by the RPMB state (unless we want to brick boards :) ) - once closed, RPMB will no longer be accesible if it was already provisioned which is not a desirable state.

Closing the shouldn't necessarily be bound to RPMB, as that is OP-TEE specific (bugs we had were because OP-TEE was burning the wrong key).

> I think this PR needs further investigation. But if you want something quick and dirty why dont you just compile it out with CONFIG_MMC?

Need sdcard support.

ldts commented:
Yes that is the case. but then lets expand there for emmc (even though it is unnecessary IMO)

Coming back to the commit itself and the points you made, well they are all wrong (sorry to be blunt but your commit statement left no choice :) ). This is not a uboot issue and the PR is not a fix, is just removing a feature.

This could be an OP-TEE issue -if OP-TEE panics instead of returning an error - which is what I suspect.

1. this uboot code does work on devices without RPMB (rpmb should not be accessible... if there is a panic, the fix is hardly removing the feature itself...I mean....). 
2. this uboot code can not write the wrong key unless there is a bug in OPTEE; and it if does, well, how is this a problem in the feature? the problem would be in optee (same than before)
3. this uboot problem doesnt cause a panic....optee might choose to panic though so back to optee.

So what this PR does is removing a feature that might be in need of OP-TEE modifications after we upgraded to 3.10 (remember this was done in 3.6)

If  OP-TEE panics on failing rpmb writes we should understand why - is it a bug, or is it required by the GP specification?
if it is required by the GP specification, then we can chose whether to carry this functionality and modify optee or not.

so well, no, there is not a bug in uboot and no, this PR is not correct. the PR should say something like remove RPMB safeguard or something like that....

ricardosalveti commented:
> Yes that is the case. but then lets expand there for emmc (even though it is unnecessary IMO)
> 
> Coming back to the commit itself and the points you made, well they are all wrong (sorry to be blunt but your commit statement left no choice :) ). This is not a uboot issue and the PR is not a fix, is just removing a feature.

Well, of course I disagree :-)

Here we're using and deciding if we should be closing or not the device based on the RPMB functionally, which also depends on the state defined by OP-TEE, which I don't think it is correct.
 
> This could be an OP-TEE issue -if OP-TEE panics instead of returning an error - which is what I suspect.
> 
> 1. this uboot code does work on devices without RPMB (rpmb should not be accessible... if there is a panic, the fix is hardly removing the feature itself...I mean....).

I don't see this as a feature, that is my point. It is confusing from my pov, as we're connecting things that shouldn't really be connected, as it is really board specific (on imx6ull the rpmb key derived from huk will be the same on open and closed devices).

Here we're making the closing process to fail if the rpmb key is not behaving as we defined, which is incorrect from my pov.

> 2. this uboot code can not write the wrong key unless there is a bug in OPTEE; and it if does, well, how is this a problem in the feature? the problem would be in optee (same than before)

Well, this same feature was created because of a bug in OP-TEE, how is this any different?

> 3. this uboot problem doesnt cause a panic....optee might choose to panic though so back to optee.
> 
> So what this PR does is removing a feature that might be in need of OP-TEE modifications after we upgraded to 3.10 (remember this was done in 3.6)

Why do we need this feature? Even on devices where we burned the incorrect RPMB key, we will still want to be able to close the device.

> If OP-TEE panics on failing rpmb writes we should understand why - is it a bug, or is it required by the GP specification?
> if it is required by the GP specification, then we can chose whether to carry this functionality and modify optee or not.
> 
> so well, no, there is not a bug in uboot and no, this PR is not correct. the PR should say something like remove RPMB safeguard or something like that....

Sure, I can update the naming, but it is already quite clear :-)

ldts commented:
um why are we not in agreement? I dont like this :)

OP-TEE is the one that knows the state of the board and the state of the RPMB. 
Sure the interface could be nicer - the current one was done as you said with RPMB in mind- but IMO it makes sense to interrogate OP-TEE before closing the board.

Then OP-TEE could check if the HUK is persistent, if RPMB is provisioned or if the board is in a secured state: the whole point of this commit was adding an extra layer of checks to prevent users from bricking boards because just checking the HAB status and fuse values is not enough IMO.

When I implemented the fuse check functionality I had the same thought in mind: making sure that we dont brick boards....I think we should continue with this vision (ultimately translates into letting the secure os validate)

ricardosalveti commented:
> um why are we not in agreement? I dont like this :)

I agree with the need for such verification, just not with the current implementation :-)

> OP-TEE is the one that knows the state of the board and the state of the RPMB.
> Sure the interface could be nicer - the current one was done as you said with RPMB in mind- but IMO it makes sense to interrogate OP-TEE before closing the board.
> 
> Then OP-TEE could check if the HUK is persistent, if RPMB is provisioned or if the board is in a secured state: the whole point of this commit was adding an extra layer of checks to prevent users from bricking boards because just checking the HAB status and fuse values is not enough IMO.

I agree that having these checks are useful and needed (to avoid bricking boards, as you said), we just need a better design then.

We could create an additional command or similar, and use it at the closing process, based on what OP-TEE exports.

Things we need to check:
- if emmc is available or not, as closing should still be allowed devices with sdcard-only
- if rpmb already has a key burned, but still allow the user to decide if he/she should proceed

This taking into account the panic issue as well (ideally without any panic during those checks).
 
> When I implemented the fuse check functionality I had the same thought in mind: making sure that we dont brick boards....I think we should continue with this vision (ultimately translates into letting the secure os validate)

Agree, we can improve the implementation for sure.

ldts commented:
To sum up, from my point of view, I dont think this PR correctly addresses the issue since it removes currently working functionality useful for imx7ulp and imx8mm; this is indirectly HUK related since the SNVS security state is checked by RPMB initt and so on...

We could carry the patch on the customer's repo, or we could try to understand why OP-TEE panics instead of returning an error (which  probably is all we need)

Or if  we decide we no longer need this functionality  sure lets remove it but I still think we should work on something along these same lines.

ricardosalveti committed:
[FIO toup] arm: dts: imx6ull: add rng

Add node for the RNGB block, as done by commit 92afd5786d9 on kernel
upstream (5.9).

Upstreaming work is needed to sync the dtsi from latest kernel.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ricardosalveti committed:
[FIO toup] arm: dts: imx6ull-14x14-evk-u-boot add rngb

Add rngb node so it can be made available to optee-os.

Signed-off-by: Ricardo Salveti <ricardo@foundries.io>

ricardosalveti commented:
Argh, reused the same branch and it assumed a merge, which is not the case here.
